File: stdlib/kpkg/kpkg.py
# test674 : stdlib5.kpkg

import os
import shutil
import time
import hashlib

import kdb
import ksc
import ksignhy
import kzip

class toolbox: # use ./temp674/
    def __init__(self):
        self.public = "" # RSA public key, empty if not using
        self.private = "" # RSA private key, empty if not using
        self.osnum = 0 # any : 0,  windows : 1, linux mint : 2
        self.name = "" # package name
        self.version = 0.0 # package version
        self.text = "" # package explanation
        self.rel_date = "" # package release date
        self.dwn_date = "" # package download date

    # pack package by osnum/dirpaths, gen file at respath
    def pack(self, osnums, dirpaths, respath):
        # setting required : public, private, name, version, text
        if len(osnums) != len(dirpaths):
            raise Exception("invalid package order")
        for i in dirpaths:
            if not os.path.isdir(i):
                raise Exception("invalid package order")
        temp = time.localtime( time.time() )
        self.rel_date = time.strftime('%Y%m%d' , temp)

        # chunk0
        worker = kdb.toolbox()
        temp = ""
        for i in range( 0, len(osnums) ):
            temp = temp + f"pkg{i} = {osnums[i]}\n"
        worker.read(temp + "name = 0\nversion = 0\ntext = 0\nrelease = 0")
        worker.fix("name", self.name)
        worker.fix("version", self.version)
        worker.fix("text", self.text)
        worker.fix("release", self.rel_date)
        chunk0 = bytes(worker.write(), encoding="utf-8") # chunk0 : basic info

        # init temp dir, pack with kzip
        dirinit(True)
        for i in range( 0, len(osnums) ):
            kzip.dozip([ dirpaths[i] ], "webp", f"./temp674/{i}.webp")

        if self.private == "":
            # not using sign
            chunk1 = b"" # chunk1 : sign info
        else:
            # using sign
            worker = kdb.toolbox()
            temp = ""
            for i in range( 0, len(osnums) ):
                temp = temp + f"pkg{i} = 0\n"
            temp = temp + "public = 0"
            worker.read(temp)
            signtool = ksignhy.toolbox()
            for i in range( 0, len(osnums) ):
                with open(f"./temp674/{i}.webp", "rb") as f:
                    hvalue = hashlib.sha3_512( chunk0 + f.read() ).digest()
                worker.fix( f"pkg{i}", signtool.sign(self.private, hvalue) )
            worker.fix("public", self.public)
            chunk1 = bytes(worker.write(), encoding="utf-8") # chunk1 : sign info

        worker = ksc.toolbox() # make basic structure
        worker.path = respath
        worker.prehead = genwebp()
        worker.subtype = b"KPKG"
        worker.reserved = ksc.crc32hash(chunk0) + ksc.crc32hash(chunk1)
        worker.writef()
        worker.linkf(chunk0)
        worker.linkf(chunk1)

        # add kzip files
        for i in range( 0, len(osnums) ):
            worker.addf(f"./temp674/{i}.webp")
        dirinit(True)

    # unpack package by os, update internal data, gen folder at temp674/temp/, returns dir path
    def unpack(self, path):
        # setting required : osnum
        dirinit(True)
        worker = ksc.toolbox()
        worker.predetect = True
        worker.path = path
        worker.readf()
        if worker.subtype != b"KPKG":
            raise Exception("invalid package")
        with open(path, "rb") as f:
            f.seek(worker.chunkpos[0] + 8)
            chunk0 = f.read( worker.chunksize[0] ) # basic data
            f.seek(worker.chunkpos[1] + 8)
            chunk1 = f.read( worker.chunksize[1] ) # sign data
        if worker.reserved != ksc.crc32hash(chunk0) + ksc.crc32hash(chunk1):
            raise Exception("invalid package")

        c0data = kdb.toolbox() # get info chunk
        c0data.read( str(chunk0, encoding="utf-8") )
        c1valid = False if len(chunk1) == 0 else True
        if c1valid:
            c1data = kdb.toolbox()
            c1data.read( str(chunk1, encoding="utf-8") )

        temp = time.localtime( time.time() ) # init install info
        self.name = c0data.get("name")[3]
        self.version = c0data.get("version")[3]
        self.text= c0data.get("text")[3]
        self.rel_date = c0data.get("release")[3]
        self.dwn_date = time.strftime('%Y%m%d' , temp)

        osnums = [ ] # get os numbers
        temp = 0
        while f"pkg{temp}" in c0data.name:
            osnums.append( c0data.get(f"pkg{temp}")[3] )
            temp = temp + 1
        if (0 not in osnums) and (self.osnum not in osnums):
            raise Exception("OS not support PKG")
        if 0 in osnums:
            pos = osnums.index(0)
        else:
            pos = osnums.index(self.osnum)

        # extract chunk corresponds osnum
        with open(path, "rb") as f:
            with open("./temp674/temp.webp", "wb") as t:
                f.seek(worker.chunkpos[pos + 2] + 8)
                size = worker.chunksize[pos + 2]
                for i in range(0, size // 10485760):
                    t.write( f.read(10485760) )
                t.write( f.read(size % 10485760) )

        # check sign
        self.public = ""
        if c1valid:
            with open("./temp674/temp.webp", "rb") as f:
                hvalue = hashlib.sha3_512( chunk0 + f.read() ).digest()
            enc = c1data.get(f"pkg{pos}")[3]
            self.public = c1data.get("public")[3]
            worker = ksignhy.toolbox()
            if not worker.verify(self.public, enc, hvalue):
                raise Exception("invalid sign")

        os.mkdir("./temp674/temp") # unpack package
        kzip.unzip("./temp674/temp.webp", "./temp674/temp", True)
        pkgpath = ( "./temp674/temp/" + os.listdir("./temp674/temp")[0] )
        pkgpath = os.path.abspath(pkgpath).replace("\\", "/")
        if pkgpath[-1] != "/":
            pkgpath = pkgpath + "/"
        with open(f"{pkgpath}_ST5_VERSION.txt", "w", encoding="utf-8") as f:
            worker = kdb.toolbox()
            worker.read("name = 0\nversion = 0\ntext = 0\nrelease = 0\ndownload = 0")
            worker.fix("name", self.name)
            worker.fix("version", self.version)
            worker.fix("text", self.text)
            worker.fix("release", self.rel_date)
            worker.fix("download", self.dwn_date)
            f.write( worker.write() )
        return pkgpath

def dirinit(mode): # init dir True : 674, False : 675
    name = "./temp674" if mode else "./temp675"
    if os.path.exists(name):
        shutil.rmtree(name)
    os.mkdir(name)

def genwebp(): # make kpkg base picture
    data0 = b""
    data0 = data0 + b"\x52\x49\x46\x46\x96\x02\x00\x00\x57\x45\x42\x50\x56\x50\x38\x20\x8a\x02\x00\x00\x50\x0f\x00\x9d\x01\x2a\x40\x00\x40\x00\x3e\x6d\x2e\x93\x46\xa4\x22\xa1\xa1\x24"
    data0 = data0 + b"\x0e\xd8\x80\x0d\x89\x6a\x00\xc0\xd4\x64\x41\x47\xb2\x7d\x80\x7e\x99\xf4\xd1\xef\xc4\x54\xde\xf6\xda\xbe\xaa\xf6\xd4\x79\x80\xfd\x6e\xfd\x80\xf7\x7c\xfe\xe5\xea"
    data0 = data0 + b"\x03\xa1\x9b\xd6\x03\xd0\x03\xf6\x33\xd3\x2f\xd9\x5b\xfc\xc5\x7d\x1f\x7a\x48\x60\xcc\x35\x8f\x03\xd0\x87\x3b\x3f\x48\x11\x21\xc2\x91\x35\xa2\x51\xd4\xe0\x39\x06"
    data0 = data0 + b"\xb4\x5d\xad\x6a\x19\x4b\x7c\x03\x68\xbe\xf8\x64\x31\xe1\x61\x3a\xff\xf9\x85\x0e\x20\x00\x5a\x8b\x01\x8b\xf2\x81\x25\x30\x60\x00\x44\x46\x4c\x57\x4c\x06\x33\x55"
    data0 = data0 + b"\xe2\x33\x87\xef\x4a\xe2\xfe\x0c\xfd\xcd\x04\xb2\x22\xef\xc0\xb5\x7f\xf0\xa7\x0b\x36\xcb\xb0\x88\x8b\xe3\x09\x14\x21\x1f\x76\xe0\x5c\xd9\x8f\xd1\xff\xfd\x6c\xfb"
    data0 = data0 + b"\xdd\xd9\x92\xaf\xad\xd1\x9a\x53\x77\x0f\xdf\xff\xbf\xb3\x79\xcb\x30\x2a\xbe\xbc\xdf\x4e\x35\x08\xa9\xe2\x80\xcd\xcc\x8c\xd6\x0f\x3a\x91\x1b\x0a\xef\x85\x55\x07"
    data0 = data0 + b"\x08\x4d\x6c\x67\x58\x7e\xa9\xc9\xa0\xb9\x61\xf7\x11\x68\xdf\x5c\x80\xa6\x19\xfe\x86\x35\x9a\xde\x4f\x75\xee\x6d\xa1\xcd\x9f\x23\x79\xb6\x2d\x10\x05\x1d\x14\x98"
    data0 = data0 + b"\xc7\x06\xfc\x33\x59\x24\x79\x13\x25\x60\x04\x4c\x18\xbb\x22\xc5\xa4\xaf\x86\xe2\x88\x03\x5a\xe9\xa4\xc0\xd6\x7d\x86\x45\xa3\xf6\x39\x34\x67\x9e\xce\x7e\x72\x28"
    data0 = data0 + b"\x51\xc0\x7d\xa0\x78\x17\xcb\x29\xc5\x5d\xf4\x5b\x06\xaf\x5a\xae\xe4\x4a\x98\xdd\xe5\x91\xca\xd2\x6b\x7b\x14\x2d\xe5\xa2\xaf\x5d\xa3\x90\xff\x69\xef\x07\x3e\x0c"
    data0 = data0 + b"\x95\xd3\x23\x6c\xef\x45\xf0\x30\x34\x48\xcc\xaf\x29\x19\x12\x3f\x6d\x61\x89\x22\xcc\xb4\x22\x34\x2b\x5f\xdb\xc1\xf6\xe6\x1d\x17\x73\xfd\x38\x84\xe9\xfc\x47\xd0"
    data0 = data0 + b"\x8c\x02\xfd\x30\xca\x65\xc8\xb6\xfa\x21\xa2\xf3\x0e\x1f\x69\x3c\x40\x7f\xcd\x9e\x49\x61\xfa\xbd\x1d\x26\xfb\x62\xe8\x87\x3f\x2b\x8c\x96\x72\x2a\x9f\xc3\xd5\x2e"
    data0 = data0 + b"\xea\x40\xb1\x9f\x32\xb4\x8a\x20\xfe\xea\xf5\xf5\x53\xd2\xdc\xe1\xe9\x98\xe8\x4f\xd2\xcc\x2a\x9f\xc2\x55\x44\xc1\x60\x06\xf1\x59\x0c\x2e\xba\xbe\x7c\x5f\x38\xa3"
    data0 = data0 + b"\x79\x76\xac\xfc\xd9\x19\x0f\xc7\xd8\xff\xfa\xb5\x23\x5c\x2d\xe5\xae\x97\xbc\x1a\xe3\xd2\xcd\x01\x79\xec\xbc\x09\x8e\xc6\x6d\x8d\xc8\x2b\x94\xbc\x80\x13\xb8\x7f"
    data0 = data0 + b"\x1b\xd2\xde\x01\x27\xcb\xc2\xfd\xdf\xc5\xce\x12\x8c\x82\x95\x33\x37\x56\x8a\x12\xf0\xcd\x98\x6c\x18\xe2\xd0\x9d\x0c\x67\xf9\x83\x49\xb9\xb3\x20\xca\xe8\x03\x56"
    data0 = data0 + b"\x6c\x21\x64\xaf\x25\xd2\xc9\x38\x00\x65\x0a\xc1\x90\x7e\x8f\x6b\x26\xef\x65\xfc\x7f\x30\xfb\xca\xde\xc5\xb1\x06\x23\x74\xb8\xb6\x80\xed\x73\x0d\x55\x16\x79\x56"
    data0 = data0 + b"\x93\x9f\x64\x2b\x98\xa4\xa0\xca\x3e\x16\xcd\x7e\x99\xf2\xd3\x31\x5e\x25\x4a\xff\x56\x27\x23\x38\x89\x92\xd8\x62\x04\x34\x37\x19\x15\x7b\x75\x13\x26\xfd\x65\xfe"
    data0 = data0 + b"\x39\xec\xac\x77\x11\x5a\x24\xbb\x69\x3d\x7f\x24\x7f\xa8\xee\x90\xc7\x22\xa2\x38\xfc\x77\xd9\x4e\xcf\xe1\x83\x40\x00\x00"
    return data0 + ( 1024 - len(data0) ) * b"\x00"

File: stdlib/kpkg/kpkg.go
// test675 : stdlib5.kpkg

package kpkg

// go get "golang.org/x/crypto/sha3"

import (
	"errors"
	"fmt"
	"os"
	"slices"
	"stdlib5/kdb"
	"stdlib5/kio"
	"stdlib5/ksc"
	"stdlib5/ksign"
	"stdlib5/kzip"
	"time"

	"golang.org/x/crypto/sha3"
)

// init toolbox, osnum 0:any 1:windows 2:linuxmint
func Initkpkg(osnum int) toolbox {
	var out toolbox
	out.Osnum = osnum
	return out
}

type toolbox struct {
	Public   string  // RSA public key, empty if not using
	Private  string  // RSA private key, empty if not using
	Name     string  // package name
	Version  float64 // package version
	Text     string  // package explanation
	Rel_date string  // package release date
	Dwn_date string  // package download date
	Osnum    int     // any : 0,  windows : 1, linux mint : 2
}

// pack package by osnum/dirpaths, gen file at respath
func (tbox *toolbox) Pack(osnums []int, dirpaths []string, respath string) error {
	// setting required : public, private, name, version, text
	if len(osnums) != len(dirpaths) {
		return errors.New("invalid package order")
	}
	tbox.Rel_date = time.Now().Format("20060102")

	// chunk0
	ts := ""
	for i, r := range osnums {
		ts = ts + fmt.Sprintf("pkg%d = %d\n", i, r)
	}
	w0 := kdb.Initkdb()
	w0.Read(ts + "name = 0\nversion = 0\ntext = 0\nrelease = 0")
	w0.Fix("name", tbox.Name)
	w0.Fix("version", tbox.Version)
	w0.Fix("text", tbox.Text)
	w0.Fix("release", tbox.Rel_date)
	ts, err := w0.Write()
	if err != nil {
		return err
	}
	chunk0 := []byte(ts) // chunk0 : basic info

	// init temp dir, pack with kzip
	dirinit(false)
	for i, r := range dirpaths {
		err = kzip.Dozip([]string{r}, "webp", fmt.Sprintf("./temp675/%d.webp", i))
		if err != nil {
			return err
		}
	}

	var chunk1 []byte
	if tbox.Private != "" {
		w1 := kdb.Initkdb()
		ts = ""
		for i := range osnums {
			ts = ts + fmt.Sprintf("pkg%d = 0\n", i)
		}
		ts = ts + "public = 0"
		w1.Read(ts)

		for i := range osnums {
			f, _ := kio.Open(fmt.Sprintf("./temp675/%d.webp", i), "r")
			data, _ := kio.Read(f, -1)
			f.Close()
			h := sha3.New512()
			h.Write(append(append(make([]byte, 0), chunk0...), data...))
			hvalue := h.Sum(nil)
			enc, err := ksign.Sign(tbox.Private, hvalue)
			if err != nil {
				return errors.New("invalid signkey")
			}
			w1.Fix(fmt.Sprintf("pkg%d", i), enc)
		}
		w1.Fix("public", tbox.Public)
		ts, err = w1.Write()
		if err != nil {
			return err
		}
		chunk1 = []byte(ts)
	}

	// make basic structure
	w2 := ksc.Initksc()
	w2.Path = respath
	w2.Prehead = genwebp()
	w2.Subtype = []byte("KPKG")
	w2.Reserved = append(ksc.Crc32hash(chunk0), ksc.Crc32hash(chunk1)...)
	w2.Writef()
	w2.Linkf(chunk0)
	w2.Linkf(chunk1)

	// add kzip files
	for i := range osnums {
		err = w2.Addf(fmt.Sprintf("./temp675/%d.webp", i))
		if err != nil {
			return err
		}
	}
	dirinit(false)
	return nil
}

// unpack package by os, update internal data, gen folder at temp675/temp/, returns dir path
func (tbox *toolbox) Unpack(path string) (string, error) {
	// setting required : osnum
	dirinit(false)
	w0 := ksc.Initksc()
	w0.Predetect = true
	w0.Path = path
	w0.Readf()
	if !kio.Bequal(w0.Subtype, []byte("KPKG")) {
		return "", errors.New("invalid package")
	}
	f, _ := kio.Open(path, "r")
	f.Seek(int64(w0.Chunkpos[0]+8), 0)
	chunk0 := make([]byte, w0.Chunksize[0])
	f.Read(chunk0) // basic data
	f.Seek(int64(w0.Chunkpos[1]+8), 0)
	chunk1 := make([]byte, w0.Chunksize[1])
	f.Read(chunk1) // sign data
	f.Close()
	if !kio.Bequal(w0.Reserved, append(ksc.Crc32hash(chunk0), ksc.Crc32hash(chunk1)...)) {
		return "", errors.New("invalid package")
	}

	c0data := kdb.Initkdb() // get info chunk
	c0data.Read(string(chunk0))
	c1valid := true
	if len(chunk1) == 0 {
		c1valid = false
	}
	c1data := kdb.Initkdb()
	if c1valid {
		c1data.Read(string(chunk1))
	}

	tv, _ := c0data.Get("name")
	tbox.Name = tv.Dat6
	tv, _ = c0data.Get("version")
	tbox.Version = tv.Dat3
	tv, _ = c0data.Get("text")
	tbox.Text = tv.Dat6
	tv, _ = c0data.Get("release")
	tbox.Rel_date = tv.Dat6
	tbox.Dwn_date = time.Now().Format("20060102")

	// get os numbers
	osnums := make([]int, 0)
	ti := 0
	tv, ta := c0data.Get(fmt.Sprintf("pkg%d", ti))
	for len(ta) != 0 {
		osnums = append(osnums, tv.Dat2)
		ti = ti + 1
		tv, ta = c0data.Get(fmt.Sprintf("pkg%d", ti))
	}
	if !slices.Contains(osnums, 0) && !slices.Contains(osnums, tbox.Osnum) {
		return "", errors.New("OSnotsupportPKG")
	}
	var pos int
	if slices.Contains(osnums, 0) {
		pos = slices.Index(osnums, 0)
	} else {
		pos = slices.Index(osnums, tbox.Osnum)
	}

	// extract chunk corresponds osnum
	f, _ = kio.Open(path, "r")
	t, _ := kio.Open("./temp675/temp.webp", "w")
	f.Seek(int64(w0.Chunkpos[pos+2]+8), 0)
	size := w0.Chunksize[pos+2]
	tb := make([]byte, 10485760)
	for i := 0; i < size/10485760; i++ {
		f.Read(tb)
		t.Write(tb)
	}
	if size%10485760 != 0 {
		tb = make([]byte, size%10485760)
		f.Read(tb)
		t.Write(tb)
	}
	t.Close()
	f.Close()

	// check sign
	tbox.Public = ""
	if c1valid {
		f, _ := kio.Open("./temp675/temp.webp", "r")
		tb, _ = kio.Read(f, -1)
		f.Close()
		h := sha3.New512()
		h.Write(append(append(make([]byte, 0), chunk0...), tb...))
		hvalue := h.Sum(nil)

		tv, _ = c1data.Get(fmt.Sprintf("pkg%d", pos))
		enc := tv.Dat5
		tv, _ = c1data.Get("public")
		tbox.Public = tv.Dat6
		flag, _ := ksign.Verify(tbox.Public, enc, hvalue)
		if !flag {
			return "", errors.New("invalid sign")
		}
	}

	// unpack package
	os.Mkdir("./temp675/temp", os.ModePerm)
	err := kzip.Unzip("./temp675/temp.webp", "./temp675/temp", true)
	if err != nil {
		return "", err
	}
	tl, _ := os.ReadDir("./temp675/temp")
	pkgpath := kio.Abs("./temp675/temp/" + tl[0].Name())
	f, _ = kio.Open(pkgpath+"_ST5_VERSION.txt", "w")
	w1 := kdb.Initkdb()
	w1.Read("name = 0\nversion = 0\ntext = 0\nrelease = 0\ndownload = 0")
	w1.Fix("name", tbox.Name)
	w1.Fix("version", tbox.Version)
	w1.Fix("text", tbox.Text)
	w1.Fix("release", tbox.Rel_date)
	w1.Fix("download", tbox.Dwn_date)
	ts, _ := w1.Write()
	kio.Write(f, []byte(ts))
	f.Close()
	return pkgpath, nil
}

// init dir True : 674, False : 675
func dirinit(mode bool) {
	var name string
	if mode {
		name = "./temp674"
	} else {
		name = "./temp675"
	}
	os.RemoveAll(name)
	os.Mkdir(name, os.ModePerm)
}

func genwebp() []byte {
	var temp []byte
	temp = append(temp, 82, 73, 70, 70, 150, 2, 0, 0, 87, 69, 66, 80, 86, 80, 56, 32, 138, 2, 0, 0, 80, 15, 0, 157, 1, 42, 64, 0, 64, 0, 62, 109, 46, 147, 70, 164, 34, 161, 161, 36)
	temp = append(temp, 14, 216, 128, 13, 137, 106, 0, 192, 212, 100, 65, 71, 178, 125, 128, 126, 153, 244, 209, 239, 196, 84, 222, 246, 218, 190, 170, 246, 212, 121, 128, 253, 110, 253, 128, 247, 124, 254, 229, 234)
	temp = append(temp, 3, 161, 155, 214, 3, 208, 3, 246, 51, 211, 47, 217, 91, 252, 197, 125, 31, 122, 72, 96, 204, 53, 143, 3, 208, 135, 59, 63, 72, 17, 33, 194, 145, 53, 162, 81, 212, 224, 57, 6)
	temp = append(temp, 180, 93, 173, 106, 25, 75, 124, 3, 104, 190, 248, 100, 49, 225, 97, 58, 255, 249, 133, 14, 32, 0, 90, 139, 1, 139, 242, 129, 37, 48, 96, 0, 68, 70, 76, 87, 76, 6, 51, 85)
	temp = append(temp, 226, 51, 135, 239, 74, 226, 254, 12, 253, 205, 4, 178, 34, 239, 192, 181, 127, 240, 167, 11, 54, 203, 176, 136, 139, 227, 9, 20, 33, 31, 118, 224, 92, 217, 143, 209, 255, 253, 108, 251)
	temp = append(temp, 221, 217, 146, 175, 173, 209, 154, 83, 119, 15, 223, 255, 191, 179, 121, 203, 48, 42, 190, 188, 223, 78, 53, 8, 169, 226, 128, 205, 204, 140, 214, 15, 58, 145, 27, 10, 239, 133, 85, 7)
	temp = append(temp, 8, 77, 108, 103, 88, 126, 169, 201, 160, 185, 97, 247, 17, 104, 223, 92, 128, 166, 25, 254, 134, 53, 154, 222, 79, 117, 238, 109, 161, 205, 159, 35, 121, 182, 45, 16, 5, 29, 20, 152)
	temp = append(temp, 199, 6, 252, 51, 89, 36, 121, 19, 37, 96, 4, 76, 24, 187, 34, 197, 164, 175, 134, 226, 136, 3, 90, 233, 164, 192, 214, 125, 134, 69, 163, 246, 57, 52, 103, 158, 206, 126, 114, 40)
	temp = append(temp, 81, 192, 125, 160, 120, 23, 203, 41, 197, 93, 244, 91, 6, 175, 90, 174, 228, 74, 152, 221, 229, 145, 202, 210, 107, 123, 20, 45, 229, 162, 175, 93, 163, 144, 255, 105, 239, 7, 62, 12)
	temp = append(temp, 149, 211, 35, 108, 239, 69, 240, 48, 52, 72, 204, 175, 41, 25, 18, 63, 109, 97, 137, 34, 204, 180, 34, 52, 43, 95, 219, 193, 246, 230, 29, 23, 115, 253, 56, 132, 233, 252, 71, 208)
	temp = append(temp, 140, 2, 253, 48, 202, 101, 200, 182, 250, 33, 162, 243, 14, 31, 105, 60, 64, 127, 205, 158, 73, 97, 250, 189, 29, 38, 251, 98, 232, 135, 63, 43, 140, 150, 114, 42, 159, 195, 213, 46)
	temp = append(temp, 234, 64, 177, 159, 50, 180, 138, 32, 254, 234, 245, 245, 83, 210, 220, 225, 233, 152, 232, 79, 210, 204, 42, 159, 194, 85, 68, 193, 96, 6, 241, 89, 12, 46, 186, 190, 124, 95, 56, 163)
	temp = append(temp, 121, 118, 172, 252, 217, 25, 15, 199, 216, 255, 250, 181, 35, 92, 45, 229, 174, 151, 188, 26, 227, 210, 205, 1, 121, 236, 188, 9, 142, 198, 109, 141, 200, 43, 148, 188, 128, 19, 184, 127)
	temp = append(temp, 27, 210, 222, 1, 39, 203, 194, 253, 223, 197, 206, 18, 140, 130, 149, 51, 55, 86, 138, 18, 240, 205, 152, 108, 24, 226, 208, 157, 12, 103, 249, 131, 73, 185, 179, 32, 202, 232, 3, 86)
	temp = append(temp, 108, 33, 100, 175, 37, 210, 201, 56, 0, 101, 10, 193, 144, 126, 143, 107, 38, 239, 101, 252, 127, 48, 251, 202, 222, 197, 177, 6, 35, 116, 184, 182, 128, 237, 115, 13, 85, 22, 121, 86)
	temp = append(temp, 147, 159, 100, 43, 152, 164, 160, 202, 62, 22, 205, 126, 153, 242, 211, 49, 94, 37, 74, 255, 86, 39, 35, 56, 137, 146, 216, 98, 4, 52, 55, 25, 21, 123, 117, 19, 38, 253, 101, 254)
	temp = append(temp, 57, 236, 172, 119, 17, 90, 36, 187, 105, 61, 127, 36, 127, 168, 238, 144, 199, 34, 162, 56, 252, 119, 217, 78, 207, 225, 131, 64, 0, 0)
	temp = append(temp, make([]byte, 1024-len(temp))...)
	return temp
}
